File: cpp/091-100/Decode Ways.cpp
class Solution {
    
    bool valid(char ch1, char ch2)
    {
        if(ch1 == '1')
            return true;
        else if(ch1 == '2')
        {
            if(ch2 <= '6')
                return true;
            return false;
        }
        return false;
    }
    
public:
    int numDecodings(string s) {
        
        int n = s.length();
        if(n == 0) return 0;
        
        int *p = new int[n+1];
        p[0] = 1;
        p[1] = 1;
        if(s[0] == '0') p[1] = 0;
        
        for(int i = 0;i < n-1;i++)
        {
            if(s[i+1] == '0')
            {
                if (s[i] >= '1' && s[i] <= '2')
                    p[i+2] = p[i];
                else 
                    p[i+2] = 0;
            } else if(s[i] == '0') {
                p[i+2] = p[i+1];
            }
            else if(valid(s[i], s[i+1]))
                p[i+2] = p[i]+p[i+1];
            else
                p[i+2] = p[i+1];
        }
        
        return p[n];
    }
};

File: python/91-100/Decode Ways.py
#!/usr/bin/env python
# coding=utf-8

class Solution:
    # @param {string} s
    # @return {integer}
    def numDecodings(self, s):
        if not s:
            return 0
        
        dp = [0]*(len(s)+1)
        dp[0] = 1
        for i in xrange(len(s)):
            if self.isValid(s[i]):
                dp[i+1] = dp[i]
            if i>0 and self.isValid(s[i-1:i+1]):
                dp[i+1] += dp[i-1]
        return dp[-1]
    
    def isValid(self,s):
        if len(s) == 1 and 0<int(s) <10:
            return True
        if len(s) == 2 and 10 <= int(s) <= 26:
            return True